GPII-413: Add a dynamic Device Reporter

javihernandez opened the pull request:
Related literature about this can be found [here], and can be tested with [this pull request to GPII/linux].

The implementation follows a similar approach to how the settingsHandlers work.
Here you can see how a solution entry should be written in order to call to a deviceReporter plugin.

<pre>
{  
    "name": "ORCA Screen Reader",
    "id": "org.gnome.orca",
    "contexts": {
    "OS": [{
        "id": "linux",
        "version": ">=2.6.26"
    }],
    "isInstalled": [
        { 
            "type": "gpii.packageKit.find",
            "name": "orca"
        }
    ]
}
</pre>

For this first example I have used the packageKit module for GNU/Linux  but we could create a few common deviceReporter plugins such as:
* A DR plugin that could look for a file in the filesystem, ie, look for /usr/opt/bin/google-chrome
* A DR plugin that could check if a process is running on the system, ie: firefox

Inspired by [this], and as a replacement for the Linux' packageKit module in Windows, a node module can be built in order to poke the Windows registry for installed solutions.

Any testing/feedback will be awesome :)
Cheers!

  [here]: http://issues.gpii.net/browse/GPII-413
  [this pull request to GPII/linux]: https://github.com/GPII/linux/pull/49
  [this]: http://stackoverflow.com/a/1458550

kaspermarkus commented:
Should this go in now, or should we wait for the updated solutionsregistry/lifecycle manager work?

klown commented:
@kaspermarkus Another factor:  there is the work I did to merge in the fire ball changes.  There is a pull request for @javihernandez's device-reporter-draft2 branch:
https://github.com/javihernandez/universal/pull/8

amb26 commented:
We need to restart discussions about this in tomorrow's arch meeting - let's put it on the agenda

michelled commented:
The new solutions entry format was discussed and @javihernandez is working on fixing the test cases.

klown commented:
@kaspermarkus, where is the new solutions entry format defined?  Is there any code on github I can look at or follow?

Thanks!

javihernandez commented:
@klown, right now (and AFAIK) the new format is [here](http://piratepad.net/ep/pad/view/c4a-arch-29-10-2014/UK4CwimNsN)

javihernandez commented:
@amb26, @kaspermarkus, @klown,

I've just updated this. As a resume of the last changes, I've added the ability to use the dynamic device reporter when running tests. The acceptance tests will use the real dynamic device reporter plugin (ie: gpii.packageKit) and the integration tests will mock these plugins by using static data from our tests definitions (see example [here](https://github.com/javihernandez/universal/commit/0e5d9cbb948cd779508eb709b1f1b005432edff1#diff-879ca59f23ea822f48dee7b0e51e1beeR58)).

klown commented:
@javihernandez,
> @klown, right now (and AFAIK) the new format is here.

Thank Javi!  Yes, I helped write that piratepad.  What I was asking @kaspermarkus was if that had been formalized somewhere in actual code yet.

I guess not (?)

javihernandez commented:
@kaspermarkus, @michelled, 

just updated it according to the new format.

javihernandez commented:
@kaspermarkus, @amb26, 

up!

amb26 commented on gpii/node_modules/deviceReporter/src/DeviceGet.js:
```diff
@@ -22,18 +22,45 @@
-            requestProxy.events.onSuccess.fire({
-                solutions: solutions,
-                OS: platformReporter.reportPlatform()
+    gpii.handleDeviceGet = function (requestProxy, installedSolutionsDataSource, platformReporter, solutionsRegistryDataSource) {
```
This isn't the right style for this integration - rather than a "FAT API" that accepts all possible arguments, we need to position the "live" device reporter as a separate grade configured at the gpii.deviceReporter level. This will replace the "installedSolutionsDataSource" component completely, rather than adding to its configuration. What I suggest is that we take the existing grade gpii.deviceReporter and rename it as gpii.deviceReporter.static, and then all the configuration you require (including the extra solutionsRegistry) can be placed into gpii.deviceReporter.live. 
Any material which is common between gpii.deviceReporter.live and gpii.deviceReporter.static can go into gpii.deviceReporter.base

Other notes: 

i) the "kettle.dataSource.simpleCallbackWrapper" was eliminated from the framework some revisions ago and can be omitted.
ii) you shouldn't duplicate the entire solutionsRegistryDataSource in any case, but you can just refer to the existing one attached to the flowManager using the reference {solutionsRegistryDataSource}. You should factor out all the code in your "else" block below into its own global function implementing gpii.deviceReporter.live.get and then just supply the "{solutionsRegistryDataSource}" in the args to its invoker.
iii) by factoring these two implementations out, you can then remove the duplication whereby each one makes a call to a dataSource and then makes a fire to requestProxy.events.onSuccess - this can all go up into the base grade.

javihernandez replied:
@amb26, splitted into 'static' and 'live' [here](https://github.com/javihernandez/universal/commit/c8d31dd8bf246dcd1a679b14d93e9a93b06e901f), but I still need to address the "Other notes".

ii) how can I use the flowManager's solutionsRegistryDataSource? By having replaced the current {solutionsRegistryDataSource} with {flowManager}.solutionsRegistryDataSource in my argument list didn't do the trick?
iii) How? Should I wrap this into a callback for deviceReporter.get? Or should I add a new invoker in gpii.deviceReporter.base?

amb26 commented on gpii/node_modules/testing/src/Integration.js:
```diff
@@ -369,3 +372,57 @@ fluid.defaults("gpii.test.integration.mockSettingsHandlerRegistry.android", {
+gpii.test.integration.mockDeviceReportersRegistry.populate = function (that) {
+    fluid.each(that.options.deviceReporters, function (options, key) {
+        var expectInstalled = [];
+        fluid.each(gpii.tests, function (tests) {
```
I'm astonished that this works at all! In general we can't assume that everything in gpii.tests.* contains a definition of the expected form, this is a dependency time bomb waiting to go off. You should instead add a specific grade name, for example, gpii.test.deviceReporterAware test into the "gradeNames" entry of your test definitions, and then insert this code into an "onCreate" listener of this grade.

javihernandez replied:
@amb26,

in response to this comment, I've made a few changes in https://github.com/javihernandez/universal/commit/b9b8ee99edaf8a30078e5928a980167c1e02e5b9

Not sure if I've fixed the race condition you mention before, but at least, it works as expected, including the tests.

javihernandez commented:
@amb26, 

ii) I'm getting an empty response from the solutions registry, so when calling to localhost:8081/device I'm getting 

```
{"solutions":[],"OS":{"id":"linux","version":"3.11.10-21-desktop"}}
```

iii) I did this in https://github.com/javihernandez/universal/commit/f842a4035e2c0b30dc6ca496d2788b7e973eb769, is this correct?

amb26 commented:
ii) That seems very odd - could you check in a branch that shows this behaviour and tell me how to trigger it? We should make sure that this is working

iii) looks good, yes

javihernandez commented:
wrt ii), I've just pushed it [here](https://github.com/javihernandez/universal/tree/tempBranch), and the commit that makes the change is [this one](https://github.com/javihernandez/universal/commit/60b1cbd9a37c0f298e71ffda2c68cd45e2094a18).

klown commented:
@javihernandez I'm not sure if this addresses the issue, but there is a problem in DeviceGet.js.  There is "func" property in the invoker section, but it should be "funcName".  See:

https://github.com/javihernandez/universal/blob/device-reporter-draft-2/gpii/node_modules/deviceReporter/src/DeviceGet.js#L28

amb26 commented:
@klown - "func" is correct here, although with the current framework there is no difference.
@javihernandez - I tried out your "tempBranch" and it works fine - a call to localhost:8080/device produces the following output:
http://pastie.org/10208472

Perhaps you could put a log statement in line 314 of Kettle's dataSource.js to read the following, and see if it helps to see what is happening:

    console.log("Calling ", func, " for dataSource get of url " + url);

klown commented:
@javihernandez, @amb26, I recently pushed a version of the ProcessReporter[1] that uses grades instead of the fat API, and it works in both development and production configurations.  I decided to come back to the device reporter and see if I could detect a difference to explain your problem @javihernandez .  But, when I run this "draft2" branch of the deviceReporter in development and production configurations, I get what is expected:
http://pastie.org/10224046 -- look for headings "Development" and "Production".

[1] https://github.com/GPII/universal/pull/355, specifically https://github.com/klown/universal/commit/8381ec7d75314561a60387af1681ca276d0f9ec0

klown commented:
@javihernandez, @amb26, never mind me.  I just read the comment about duplicating "the entire solutionsRegistryDataSource", and realized that's what the issue is.  I'll look into it since the ProcessReporter is likewise duplicating.

klown commented:
@javihernandez , @amb26, When I test Javi's "tempBranch", I get the same result as he does.  It depends on running with the development configuration, i.e. using [all.development.dr.production.json](
https://github.com/javihernandez/universal/blob/tempBranch/gpii/configs/all.development.dr.production.json)

Adding a "console.log()" as @amb26 suggested gives the following.  The problem is the url to the test data ends in "null.json" instead of "linux.json".  Trying to figure out why:
```
Calling  function (fileName, options, data) {
    var promise = fluid.promise(),
        method = "readFile",
        fsCallback = function (error, data) {
            if (error) {
                promise.reject({
                    isError: true,
                    message: error.message
                });
            } else {
                promise.resolve(options.operation === "set" ? undefined : data);
            }
        },
        args = [fileName, "utf8"];
    if (options.operation === "set") {
        method = "writeFile";
        args.splice(1, 0, data);
    } else {
        if (!fs.existsSync(fileName)) {
            promise.resolve(undefined); // dataSource semantics - read from nonexistent file produces undefined
            return;
        }
    }
    args.push(kettle.wrapCallback(fsCallback));
    fs[method].apply(null, args);
    return promise;
}  for dataSource get of url /home/clown/Development/GPII/DeviceReporter/node_modules/universal/gpii/node_modules/flowManager/../../../testData/solutions/null.json
```

klown commented:
@javihernandez, @amb26,

I wrote:
> The problem is the url to the test data ends in "null.json" instead of "linux.json". Trying to figure out why:

The first argument to solutionsRegistryDataSource() at [line 100](https://github.com/javihernandez/universal/blob/tempBranch/gpii/node_modules/deviceReporter/src/DeviceReporter.js#L100) is 'null.  But it needs to communicate the OS somehow.

I don't know how to do that.  To test the hypothesis, I tried:
```
var pltfrm = { os: deviceReporter.platformReporter.reportPlatform().id };
solutionsRegistryDataSource.get(pltfrm, function onSuccess(entries) { ...
```
And that worked.  But, there's got to be a better way.

javihernandez commented:
@klown, 

that's correct, I was missing that. Thanks a lot!!
I'm updating it now.

javihernandez commented:
up

amb26 commented on gpii/node_modules/testing/src/Integration.js:
```diff
@@ -371,3 +371,60 @@ fluid.defaults("gpii.test.integration.mockSettingsHandlerRegistry.android", {
+
+gpii.test.integration.deviceReporterAware.populate = function (that) {
+    fluid.each(that.options.deviceReporters, function (options, key) {
+        if (fluid.getGlobalValue(key)) {
```
If a 2nd onCreate fires, one should imagine that it is part of a further test case which may well require different expectInstalled contents. So we should just remove this branch.

javihernandez replied:
This is how I'm checking whether we need to mock or not. Isn't it the right place/way for this?

amb26 replied:
The nameResolver should give a completely fresh name for the deviceReporter under the mock configuration - for example, the way our settingsHandlers do. There should never be a question of overwriting the global name of the real deviceReporter, etc.

amb26 replied:
http://www.w3.org/Provider/Style/URI.html

amb26 commented on gpii/node_modules/testing/src/Integration.js:
```diff
@@ -371,3 +371,60 @@ fluid.defaults("gpii.test.integration.mockSettingsHandlerRegistry.android", {
+            return;
+        }
+        var expectInstalled = [];
+        fluid.each(gpii.tests.deviceReporterAware, function (tests) {
```
This is still bizarre - why are we looping over all of the global tests, rather than the options for the test which is currently running?

javihernandez replied:
Yeah, my fault, I introduced this for debugging a problem and after that I forgot removing it.

javihernandez commented:
@amb26 

I've updated this, but I'm not feeling specially proud of it. Can you take a look at it?

amb26 commented on gpii/node_modules/deviceReporter/src/DeviceReporter.js:
```diff
@@ -32,21 +32,85 @@ fluid.defaults("gpii.deviceReporter", {
         },
-        platformReporter: {
-            type: "gpii.platformReporter.native"
+        resolveName: {
```
This method is ok but messy - it will be clearer for readers (and for YOU, VANTTUNEZZ!) if you just use the same scheme that we have for settings handlers - that is, just inject the nameResolver as a component of gpii.deviceReporter.base and expect for all users to invoke it from there (rather than make a fresh invoker) - that is. deviceReporter.nameResolver.resolveName

javihernandez commented:
@klown :beers: :beers: :beers: :beers:

colinbdclark commented:
:clap: woot @klown @javihernandez!

klown commented:
c’est magnifique, @javihernandez.  :notes: (in honour of tuxguitar).
